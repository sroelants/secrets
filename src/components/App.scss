@import './common';

html {
    overflow: scroll;
    overflow-x: hidden;
}

@media (max-width: $bp-small) {
    html {
        font-size: 0.7rem;
    }
}

body {
    background: $white;
    color: $black;
    font-family: $sansfont;
}

// Main [MAIN] -----------------------------------------------------------------

.footer {
    display: flex;
    align-items: center;
    justify-content: center;
    min-height: 150px;

    a {
        font-weight: 700;
        text-decoration: none;
        color: inherit;
        font-style: italic;
    }
}
