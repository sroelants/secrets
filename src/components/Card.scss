@import "./common";

.card {
  font-size: 1.5rem;
  min-height: 5rem;
  padding: 1rem;
  border: 1px solid $lgrey;
  border-radius: 2px;
  background: $white;
  box-shadow: 6px 6px 15px rgba(0, 0, 0, 0.03), 2px 4px 4px rgba(0, 0, 0, 0.05),
    0 0 10px rgba(0, 0, 0, 0.05);
  animation: slide-in;
  animation-duration: 0.3s;
  width: 90%;
  margin: 30px auto;
}

@media (min-width: $bp-medium) {
  .card {
    font-size: 1.3rem;
    width: 50%;
  }
}

@keyframes slide-in {
  from {
    opacity: 0;
    transform: translateY(100px);
  }
  to {
    opacity: 1;
    transform: translateY(0);
  }
}

.card__header {
  font-size: 0.8em;
  color: $grey;
  display: flex;
  justify-content: space-between;
}

.card__date {
  text-transform: uppercase;
}

.card__secret {
  margin-top: 10px;
  font-family: $seriffont;
  font-size: 1em;
  line-height: 1.5;
}

@media (min-width: $bp-small) {
  .card__secret {
    font-size: 0.9em;
  }
}

.card__like {
  width: 1em;
  filter: grayscale(1);
  opacity: 0.7;
  transition: all 0.3s linear;
  cursor: pointer;
  margin: 0 10px 0 10px;
  position: relative;
  top: 3px;
}

.liked {
  opacity: 1;
  filter: grayscale(0);
  animation: bounce;
  animation-duration: 0.2s;
}

@keyframes bounce {
  0% {
    transform: scale(1);
  }
  50% {
    transform: scale(0.5);
  }
  100% {
    transform: scale(1);
  }
}
