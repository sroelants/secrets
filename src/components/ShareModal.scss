@import "common";

.shareform {
  align-items: flex-end;
  display: flex;
  flex-direction: column;
  width: 30em;
}

.shareform__textarea {
  background: inherit;
  border: 1px solid $grey;
  border-radius: 4px;
  box-sizing: border-box;
  color: $black;
  font-family: $sansfont;
  font-size: 1.3rem;
  height: 300px;
  padding: 10px;
  resize: none;
  transition: border 0.2s linear;
  width: 100%;
}

.shareform__textarea:focus {
  border: 2px solid $black;
  outline: none;
}

.shareform__button {
  -webkit-appearance: none;
  font-family: $sansfont;
  margin: 10px 0 10px 10px;
  padding: 6px;
  font-size: 1.3rem;
  width: 5em;
  border: none;
  border-radius: 4px;
  outline: none;
  transition: all 0.1s linear;
  &:hover {
    cursor: pointer;
    background-color: $lgrey;
    color: $black;
  }
  &:active {
    transform: translateY(2px);
  }
}
.shareform__submit {
  background: $black;
  color: $white;
}

.shareform__clear {
  background: $grey;
  color: $white;
}

.shareform__success {
  align-items: center;
  color: inherit;
  display: flex;
  font-size: 1.3rem;
  justify-content: center;
  min-height: 7em;
}

.shareform__success-icon {
  animation-duration: 0.3s;
  animation-name: grow;
  margin: 30px;
}

@keyframes grow {
  from {
    transform: scale(0);
  }
  to {
    transform: scale(1);
  }
}
