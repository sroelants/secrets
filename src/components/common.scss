$white: #fafafa;
$black: #444;
$lblack: #555;
$grey: #888;
$lgrey: hsla(0, 0, 60, 0.3);
$sansfont: "Nunito", sans-serif;
$seriffont: "Libre Baskerville", serif;

$bp-small: 580px;
$bp-medium: 900px;
$bp-wide: 1200px;
