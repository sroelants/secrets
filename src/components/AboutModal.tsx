import './Modal.scss';
import React from 'react';
import { ModalType } from './App'

export interface IModal {
  visible: boolean;
  closeHandler: (t: ModalType | null) => void;
}

export const AboutModal: React.FC<IModal> = ({ visible, closeHandler }) => {
  return (
    <>
      <div className={visible ? "about-modal modal visible" : "about-modal modal"} >
        <button className="modal__close" onClick={() => {
          closeHandler(null);
        }}>
        <img aria-label="close"
          src="assets/img/close.svg"
          width="25px"
          alt="close" />
        </button>
        <div className="modal__title">About Secrets</div>
        <div className="modal__content">
          <p>
          We all know that feeling of relief after sharing a long-kept secret
          with someone. A weight is lifted. What seemed like a terrible or
          embarrasing thing all of a sudden doesn't seem that important anymore.
          Sharing secrets with people brings us perspective. We're not only
          opening ourselves up to them, we're opening up to ourselves.
        </p>

        <p>
          The science <a href="https://psycnet.apa.org/record/2017-20428-001"
          target="_blank" rel="noopener noreferrer"> has
          long been in </a> on the benefits of sharing secrets with
          others. Keeping secrets is a mental, emotional and physical burden.
          Our minds keep drifting and focussing on them. We are more anxious
          and adrenaline levels and blood pressure are above average.
        </p>
        <p>
          On the flip side, sharing secrets with others has a tremendous effect
          on us. Even if we're uncomfortable sharing with someone close to us,
          for fear of judgement, we ofter share our deepest secrets with people
          we're not close to, even complete strangers like a hairdresser or
          taxi driver. Sharing secrets forms bonds of trust, they strengthen
          social cohesion, even between strangers. Aside from benefiting us
          emotionally and physically, sharing secrets benefits our society as a
          hole.
        </p>
        <p>
          This page is meant as a platform for people to share their secrets
          with strangers, completely anonymously. Perhaps the person sitting
          next to you on the bus tomorrow will know and sympathize with
          something you haven't yet had the courage to face up to, and that is
          an encouraging thought.
        </p>
        </div>
      </div >
      <div className="dimmer"></div>
    </>
  );
};
