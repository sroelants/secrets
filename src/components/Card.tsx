import React, {useState } from "react";
import "./Card.scss";

export interface ICard {
  secretId: number;
  secret: string;
  date: string;
  likes: number;
}

export const Card: React.FC<ICard> = ({ secretId, secret, date, likes }) => {
  // Check if secret is already liked (in localStorage)
  const storedLike = localStorage.getItem(secretId.toString());
  let alreadyLiked: boolean;
  if (storedLike === null) {
    alreadyLiked = false;
  } else {
    alreadyLiked = true;
  }

  const [liked, setLiked] = useState(alreadyLiked);
  const [localLikes, setLocalLikes] = useState(likes);
  const dateObject = new Date(date);
  const shortDate = dateObject
    .toString()
    .split(" ")
    .slice(1, 4)
    .join(" ");

  const handleLike = () => {
    // Send put request to server.
    const update = {
      date_posted: date,
      likes: !liked ? likes + 1 : likes - 1,
      secret,
    };

    fetch("https://secrets-backend.herokuapp.com/api/secret/" + secretId, {
      body: JSON.stringify(update),
      headers: new Headers({ "Content-Type": "application/json" }),
      method: "PUT",
    });

    // Set current secret's like status in localStorage
    if (liked) {
      localStorage.removeItem(secretId.toString());
    } else {
      localStorage.setItem(secretId.toString(), "liked");
    }

    setLocalLikes(localLikes + (!liked ? 1 : -1)); // Update shown like count.
    setLiked(!liked); // Update shown like status
  };

  return (
    <div className="card">
      <div className="card__header">
        <div className="card__date">{shortDate}</div>
        <div className="card__likes">
          {localLikes}
          <img alt="like"
            className={liked ? "card__like liked" : "card__like"}
            src="/assets/img/like.svg"
            onClick={handleLike}
          />
        </div>
      </div>
      <div className="card__secret">{secret}</div>
    </div>
  );
};
