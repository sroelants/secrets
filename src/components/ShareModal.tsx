import "./Modal.scss";
import "./ShareModal.scss";
import React, { useState, useEffect } from "react";
import { ModalType } from "./App";

export interface IModal {
  visible: boolean;
  closeHandler: (t: ModalType | null) => void;
}

export const ShareModal: React.FC<IModal> = ({ visible, closeHandler }) => {
  const [secret, setSecret] = useState("");
  const [submitted, setSubmitted] = useState(false);

  // When turned visible (visible changes and is set to true)
  // reset submitted to false.
  useEffect(() => { if (visible) { setSubmitted(false); } }, [visible]);

  const handleChange = (e: any) => {
    setSecret(e.target.value);
  };

  const handleSubmit = (e: any) => {
    const content = e.target.querySelector(".shareform__textarea").value;
    const body = {
      date_posted: (new Date(Date.now())).toISOString(),
      likes: 0,
      secret: content,
    };

    fetch("https://secrets-backend.herokuapp.com/api/secrets", {
      body: JSON.stringify(body),
      headers: (new Headers({ "Content-Type": "application/json" })),
      method: "POST",
    });

    setSubmitted(true);
    e.preventDefault();
  };

  return (
    <>
      <div className={visible ? "share-modal modal visible" : "share-modal modal"} >
        <button className="modal__close" onClick={() => {
          closeHandler(null);
          setSecret("");
        }}>
<img aria-label="close"
          src="assets/img/close.svg"
          width="25px"
          alt="close" />
        </button>
        {(!submitted)
          ? (
            <>
              <div className="modal__title">Share a secret</div>
              <div className="modal__content">
                <form onSubmit={handleSubmit} className="shareform">
                  <textarea onChange={handleChange}
                    value={secret}
                    className="shareform__textarea"
                    placeholder="Share your secret." />
                  <div>
                    <button
                      type="button"
                      className="shareform__button shareform__clear"
                      onClick={() => setSecret("")}> Clear </button>
                    <input type="submit"
                      value="Share"
                      className="shareform__button shareform__submit" />
                  </div>
                </form>
              </div>
            </>)
          : (
            <div className="shareform__success">
              <img src="/assets/img/icon--checkmark.svg"
                   width="50px"
                   className="shareform__success-icon"
                   alt="Submitted" />
                Thank you for sharing your secret!
                 It should appear on the page shortly.
            </div>)}
      </div >

      <div className="dimmer"></div>
    </>
  );
};
