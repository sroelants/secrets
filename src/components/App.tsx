import "./App.scss";
import React, { useState } from "react";
import { AboutModal } from "./AboutModal";
import { Header, IHeader } from "./Header";
import { Secrets } from "./Secrets";
import { ShareModal } from "./ShareModal";

export enum ModalType {
  About,
  Share,
}

const App: React.FC = () => {
  const [modalState, setModalState] = useState<ModalType | null>(null);

  const setModalHandler = (mode: ModalType | null) => {
    setModalState(mode);
  };

  const HeaderProps: IHeader = { clickHandler: setModalHandler };

  return (
    <div className="App">
      <Header {...HeaderProps} />
      <main className="main">
        <Secrets />
      </main>
      <AboutModal visible={modalState === ModalType.About}
        closeHandler={setModalHandler} />
      <ShareModal visible={modalState === ModalType.Share}
        closeHandler={setModalHandler} />
      <footer className="footer">
        <p>
        Created by
        <a href="https://www.samroelants.com"
          target="_blank"
          rel="noopener noreferrer">
           &nbsp;Sam Roelants
        </a>
      </p>
      </footer>
    </div>
  );
};

export default App;
