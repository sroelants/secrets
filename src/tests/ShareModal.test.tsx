import React from "react";
import renderer from "react-test-renderer";
import {ShareModal} from "../components/ShareModal";
import {shallow, mount} from "enzyme";

// Unit
describe("ShareModal", () => {
  it("renders without crashing", () => {
    shallow(<ShareModal visible={true} closeHandler={() => {}} />);
  });

  it("Matches the snapshot", () => {
    const output = renderer.create(
      <ShareModal visible={true} closeHandler={() => {}} />
    );
    expect(output).toMatchSnapshot();
  });
});
