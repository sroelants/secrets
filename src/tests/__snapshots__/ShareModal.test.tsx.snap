// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShareModal Matches the snapshot 1`] = `
Array [
  <div
    className="share-modal modal visible"
  >
    <button
      className="modal__close"
      onClick={[Function]}
    >
      <img
        alt="close"
        aria-label="close"
        src="/assets/img/close.svg"
        width="25px"
      />
    </button>
    <div
      className="modal__title"
    >
      Share a secret
    </div>
    <div
      className="modal__content"
    >
      <form
        className="shareform"
        onSubmit={[Function]}
      >
        <textarea
          className="shareform__textarea"
          onChange={[Function]}
          placeholder="Share your secret."
          value=""
        />
        <div>
          <button
            className="shareform__button shareform__clear"
            onClick={[Function]}
            type="button"
          >
             Clear 
          </button>
          <input
            className="shareform__button shareform__submit"
            type="submit"
            value="Share"
          />
        </div>
      </form>
    </div>
  </div>,
  <div
    className="dimmer"
  />,
]
`;
